cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -914,9 +915,9 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
             return withKafkaDiff(kafkaSetOperations.reconcile(namespace, kafkaCluster.getName(), kafkaSs));
         }
 
-        Future<ReconciliationState> kafkaRollingUpdate() {
+        Future<ReconciliationState> kafkaRollingUpdate(Supplier<Date> dateSupplier) {
```

Comment:
Really using `Supplier` is being lazy. It has the signature we need, but the semantics of the interface (or rather, in that case, the semantics implied by its membership of `java.util.function`) are that it should be used in functional-style programming. That's not how we're using it. I think we should avoid being lazy and declare our own `Time` interface. 

I'm a fine one to lecture here, as I'm guilty of this sort of laziness all the time. Please pull be up on it in the future.

Reply:
Tbh I don't see a real reason for writing a new interface from scratch when `Supplier` provides what we need. The semantics is about supplying a `Date` which is what we need. Why should we write a new one? I didn't get your point, sorry.

Reply:
Would you argue that you're never going to declare another interface having a single no args method, simply because `Supplier` exists, and you can use that? Because logically you _could_. Likewise if you just needed an interface with a single 1-arg method there's `Function`. You could do this, and your code would work, but it wouldn't make it any simpler to understand. For a start everywhere where you used a `Supplier` you'd have to document the semantics you expected for _that particular use site_. For example, in a crypto library a `Supplier<Long>` that returns prime numbers is a very different thing than a `Supplier<Long>` that returns random numbers.

Having a dedicated interface provides somewhere to declare what the required semantics are and they allow you to leverage type safety. 

You could compare the use of `Supplier<Date>` here with the use of `Predicate<Pod>` for the rolling. I feel that `Predicate<Pod>` is _just about_ OK, because what we need really does have precisely the semantics of a predicate on `Pods`. But a `Supplier`'s Javadoc is "Represents a supplier of results.", well that doesn't really describe quite how we're using it here, does it? We're not using it in a abstract way. Supplying the current time is quite a concrete thing, so a more concrete interface is appropriate.

Reply:
I'm not saying that it doesn't work, or is not meaningful. I am saying that using `Supplier`, `Function` and friends everywhere is a slippery slope to less type-safe code. An interface is a contract, and the contract needed here is more refined than merely the supply of any old random `Date`: It is supposed to be the _current_ date. The problem isn't so much the use of `Supplier` per-se, it's more that we're moving the down a slippery slope.

Reply:
I don't see the slippery slope :-) Even `Supplier<Date>` defines a contract ... "providing a date" which is what we need and is also well defined, or?